### PR TITLE
DAOS-12212 uns: add option to link an existing container in the UNS

### PR DESCRIPTION
**Pull request by mchaarawi**
Today this is only supported on container creation. This PR adds an option to allow linking a container that already exists in the namespace.

Required-githooks: true

Signed-off-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/client/dfs/duns.c [hunk 1]
```diff
@@ -852,6 +852,51 @@ duns_create_lustre_path(daos_handle_t poh, daos_pool_info_t info, const char *pa
...
+	if (len < 0) {
+		D_ERROR("Failed to create foreign LOV/LMV value\n");
+		return EINVAL;
+	}
+
+	if (type == DAOS_PROP_CO_LAYOUT_POSIX) {
+		rc = (*dir_create_foreign)(path, mode, LU_FOREIGN_TYPE_SYMLINK, (__u32)type, str);
+		if (rc) {
+			D_ERROR("Failed to create Lustre dir '%s' with foreign "
+				"LMV '%s' (rc = %d).\n", path, str, rc);
```
Comment (daltonbohning) [inline src/client/dfs/duns.c:882]: Should we format the rc similar to this?
https://github.com/daos-stack/daos/blob/1e654aa4c1086d66db3e1339007a2ae9ab0e4575/src/client/dfs/duns.c#L570

File: src/client/dfs/duns.c [hunk 2]
```diff
@@ -1086,6 +1131,194 @@ duns_create_path(daos_handle_t poh, const char *path, struct duns_attr_t *attrp)
...
+int
+duns_link_cont(daos_handle_t poh, const char *cont, const char *path)
+{
+	daos_handle_t		coh;
+	daos_prop_t		*prop;
+	struct daos_prop_entry	*entry;
+	daos_pool_info_t	pinfo = {0};
+	daos_cont_info_t	cinfo = {0};
+	daos_cont_layout_t	type;
+	char			pool_str[37], cont_str[37];
```
Comment (daltonbohning) [inline src/client/dfs/duns.c:1143]: Is this what `DAOS_UUID_STR_SIZE` is for?
```suggestion
	char			pool_str[DAOS_UUID_STR_SIZE], cont_str[DAOS_UUID_STR_SIZE];
```

**Review by daltonbohning — APPROVED** [review]

File: src/client/dfs/duns.c (hunk 1 above)
Comment (mchaarawi) [reply src/client/dfs/duns.c:882]: this is just copy paste from above.. probably need to do a wider scan of errno rc formatting in this file and others independent of this PR.

File: src/client/dfs/duns.c (hunk 2 above)
Comment (mchaarawi) [reply src/client/dfs/duns.c:1143]: sure this looks better. ill update if re-push is needed. although not too concerning as uuids will never change length :-)

File: src/client/dfs/duns.c [hunk 3]
```diff
@@ -1086,6 +1131,194 @@ duns_create_path(daos_handle_t poh, const char *path, struct duns_attr_t *attrp)
...
+	uuid_unparse(cinfo.ci_uuid, cont_str);
+
+	if (type == DAOS_PROP_CO_LAYOUT_POSIX) {
+		struct statfs	fs;
+		char		*dir, *dirp;
+		mode_t		mode = S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH;
+		size_t		path_len;
+
+#ifdef LUSTRE_INCLUDE
+		if (fs.f_type == LL_SUPER_MAGIC) {
```
Comment (bfaccini) [inline src/client/dfs/duns.c:1195]: Arghhhhh, looks like recent commit ebc3676f2b9876896b66be15473a640805473d35 (DAOS-6540 dfuse: Enable multi-user support in dfuse (#3956)) from @ashleypittman had already removed the statfs() syscall before testing if "fs.f_type == LL_SUPER_MAGIC" in duns_create_path() so you have replicated the error by using cut/paste of this piece of code !!...

**Review by bfaccini — CHANGES_REQUESTED** [review]

File: src/client/dfs/duns.c (hunk 3 above)
Comment (mchaarawi) [reply src/client/dfs/duns.c:1195]: if not i'll just fix it with this PR

Comment (mchaarawi) [reply src/client/dfs/duns.c:1208]: just updated this PR to fix

**Review by daltonbohning — APPROVED** [review]

File: src/client/dfs/duns.c (hunk 3 above)
Comment (bfaccini) [reply src/client/dfs/duns.c:1208]: > ah ok.. so is there a PR on master that fixes that?

No I just discovered this when doing review for this PR !! Thanks to have fixed it also in duns_create_path().

**Comment by mchaarawi** [comment]
container/snapshot_aggregation: DAOS-12326

**Pull request merged**